### ISSUE TITLE
Static linking for non-Windows systems and conda environment

### ISSUE THREAD

**Issue by gioboske** [Nuitka/Nuitka#595]
A new plan on how to manage static linking on non-Windows systems was outlined by @kayhayen [here](https://github.com/Nuitka/Nuitka/pull/581#issuecomment-569212442) (PR #581).

Basically, the proposal is to make static linking and using conda-cc mandatory on non-Windows.

This issue is related to #95 (still open) and replace issue #580 (that was only related to check for the if Python 3.6+ belongs to a conda environment or not).

**Comment by kayhayen** [comment]
What I was wondering, if Anaconda environments always set the `CC` variables, and that would work then, and if anybody but me runs conda without an environment activated. Those of course always win if present.

**Comment by kayhayen** [comment]
So, this is what happened on Github Actions with Python 3.6, but I believe it's others too, but they get cancelled on first error:


> +Traceback (most recent call last):
> +  File "/Users/runner/runners/2.163.1/work/Nuitka/Nuitka/tests/basics/BuiltinsTest.py", line 29, in <module>
> +    from math import copysign
> +ImportError: dlopen(/usr/local/miniconda/lib/python3.6/lib-dynload/math.cpython-36m-darwin.so, 2): Symbol not found: _PyExc_MemoryError
> +  Referenced from: /usr/local/miniconda/lib/python3.6/lib-dynload/math.cpython-36m-darwin.so
> +  Expected in: flat namespace
> + in /usr/local/miniconda/lib/python3.6/lib-dynload/math.cpython-36m-darwin.so

It's possible that the math module is not dynamic normally for Anaconda, not being able to load PyExc_MemoryError and only that is strange. I checked its declaration and didn't see anything that made me believe it would not be exported.

Not sure how much static linking and LTO was going on, but it might have removed the external visibility of a symbol like this, but that would apply probably to others as well.

Not sure how to deal with it, disabling it again seems prudent. Debugging this remotely is not so easy for me, surely time consuming. I will have to check with Linux and Anaconda if this is happening too, then I would at least be able to look at it.

Yours,
Kay

**Comment by kayhayen** [comment]
Now that I look at it, the macOS workflow, doesn't install the condaCC at all yet, and doesn't use it, so it won't be much LTO going on, or so I suppose. Of course that might be the difference that clang of the OS would do it and gcc on Linux does not.

On 3.6 with Linux I am not getting these errors. I need to try out if that is for lack of condaCC usage, otherwise disabling it will not work. It could also be that no static libpython is installed, but that should be true for macOS and Linux similarily.

For 3.7 I get duplicate symbols from the static library, the usual one, seems with 3.7 the Anaconda got worse there too. For 3.8 on macOS I have seen it too, albeit another file.  This is all really bad.

Yours,
Kay

**Comment by kayhayen** [comment]
I am going slow on this, I am puting Linux Anaconda 3.6 through all tests first, then I will make it use its CondaCC and see if it still passes, or starts getting the issue. That way I will be on stable grounds and not waste too much time on this. Static linking to work on one platform would be good enough for me to hand the issue on macOS off to somebody else, but right now I am not even sure it's feasible.

**Comment by gioboske** [comment]
Thank you for the update @kayhayen.
I'll try to make some test in parallel on my macOS, but unfortunately I don't have much time in these days and I'd like first to deal with #562 (at least to propose a draft script).

**Comment by kayhayen** [comment]
Yes, please do that. I think I can take it from here. I am currently trying to get the debugger being used on Python automatically by `compare_with_cpython` which should make this easy and automatic to debug.

Once I have a C call stack, I am pretty optimistic, this cannot be a hard thing, but we shall see.

Yours,
Kay

**Comment by kayhayen** [comment]
Seems, there is no gdb on Github Actions with macOS, but I am guessing lldb will be there, and luckily my Debian has it, so I managed to develop it as a fallback, which is probably a good idea anyway, as clang might the only compiler installed and no gdb on other systems, e.g. some BSDs too.

**Comment by sztomi** [comment]
> the proposal is to make static linking and using conda-cc mandatory on non-Windows

I think nuitka would lose a lot of portability if the toolchain and linking style was mandated. What's the upside of doing so?

**Comment by unrelentingtech** [comment]
I'd like to just have static linking as **an option with the current setup** (no conda stuff). e.g. I already have a static libpython on FreeBSD:

```
% la /usr/local/lib/python3.9/config-3.9/libpython3.9.a
-rw-r--r--  1 root  wheel   6.2M Dec  8 15:54 /usr/local/lib/python3.9/config-3.9/libpython3.9.a
```

But even better: I'd like to be able to provide an LLVM bitcode compiled `libpython.bc` to fully LTO with libpython. This can be obtained quite easily with

```
CFLAGS+=	-flto
LDFLAGS+=	-flto -Wl,-save-temps
```

when building Python:

```
% la work/Python-3.9.4/libpython3.9.so.1.0.0.5.precodegen.bc
-rw-r--r--  1 greg  greg   6.9M May  8 16:42 work/Python-3.9.4/libpython3.9.so.1.0.0.5.precodegen.bc
```

**Comment by kayhayen** [comment]
@unrelentingtech I hear you, and there is already `Options.shallUseStaticLibPython()` to control it. I think I will add a static option to override it now, this will allow more experimentation. As for macOS and Anaconda, I might have a chance to look at that.

The reason the static linking is not done everytime it seems possible, is the incompatibility of compilers and delivered `libpython.a` with bytecode in at least Anaconda on Linux. I am not aware of the `libpython.bc` notion. Therefore beware that this code may not cut it for you:

```
    static_lib_filename = os.path.join(
        python_lib_path, "libpython" + python_abi_version + ".a"
    )
```

If possible, can you enhance the check, and point me to what this is? I have so far seen only archives with bytecode, and I believe that is supposed to be OK. What's going on with Linux e.g. is that the C compiler used crashes on the bytecode, or rejects it as too old or too new. And then I get to support confused users, if I do this by default.

**Comment by kayhayen** [comment]
There is now the option `--static` on factory, soon to be renamed to `--static-libpython` which I am further refining. It should inform in `auto` mode (which is going to be default), if it would be available. And it should error out if `yes", but I have yet to add those things, but it would work for you now.

https://nuitka.net/doc/factory.html

**Comment by kayhayen** [comment]
This should be usable now.

**Comment by kayhayen** [comment]
Part of the last pre-release.

**Comment by kayhayen** [comment]
Note that also Nuitka-Python is coming to Windows, where static linking will be baseline with Nuitka. 

But the option is part of 0.6.15 release and can be used now.
